include/nektar_interface/projection/algorithm_types.hpp

```diff
@@ -0,0 +1,269 @@
+    return 0;
+  }
+  // Need (a + 1) * local_required <= local_mem_max
+  //      a <= local_mem_max/local_requred - 1
+  uint64_t a = (local_mem_max / local_required) - 1;
+  // not sure about the sub_group_sizes thing but is the closest I can find
+  size_t min_sub_group_size =
+      dev.is_gpu()
+          ? (dev.get_info<sycl::info::device::sub_group_sizes>()[0])
+          : 1;
+  // Will return 0 if can't find something that is a multiple of
+  //(min_sub_group_size + 1) that fits
+  return ROUND_DOWN_TO_MULTIPLE(min_sub_group_size, a);
+}
+} // namespace Private
+
+struct ThreadPerCell {
+  template <int nmode, typename T, int alpha, int beta, typename Shape>
+  std::optional<sycl::event> static inline project(DeviceData<T> &data,
+                                                       int componant,
+                                                       sycl::queue &queue) {
+    sycl::range<1> range{static_cast<size_t>(data.ncells)};
+    if constexpr (Shape::dim == 2) {
+      return queue.submit([&](sycl::handler &cgh) {
+        cgh.parallel_for<>(range, [=](sycl::id<1> idx) {
+          auto cellx = data.cell_ids[idx];
+          auto npart = data.par_per_cell[cellx];
+
+          if (npart == 0)
+            return;
+
+          auto cell_dof = &data.dofs[data.dof_offsets[cellx]];
+          for (int part = 0; part < npart; ++part) {
+            auto xi0 = data.positions[cellx][0][part];
+            auto xi1 = data.positions[cellx][1][part];
+            T eta0, eta1;
+            Shape::template loc_coord_to_loc_collapsed<T>(xi0, xi1, eta0, eta1);
+            auto qoi = data.input[cellx][componant][part];
+            Shape::template project_one_particle<nmode, T, alpha, beta>(
+                eta0, eta1, qoi, cell_dof);
+          }
+        });
+      });
+    } else {
+      return queue.submit([&](sycl::handler &cgh) {
+        cgh.parallel_for<>(range, [=](sycl::id<1> idx) {
+          auto cellx = data.cell_ids[idx];
+          auto npart = data.par_per_cell[cellx];
+
+          if (npart == 0)
+            return;
+
+          auto cell_dof = &data.dofs[data.dof_offsets[cellx]];
+          for (int part = 0; part < npart; ++part) {
+            auto xi0 = data.positions[cellx][0][part];
+            auto xi1 = data.positions[cellx][1][part];
+            auto xi2 = data.positions[cellx][2][part];
+            T eta0, eta1, eta2;
+            Shape::template loc_coord_to_loc_collapsed<T>(xi0, xi1, xi2, eta0,
+                                                          eta1, eta2);
+            auto qoi = data.input[cellx][componant][part];
+            Shape::template project_one_particle<nmode, T, alpha, beta>(
+                eta0, eta1, eta2, qoi, cell_dof);
+          }
+        });
+      });
+    }
+  }
+};
+
+// Untested on Microsoft
+// c++20 has a solution for this std::source_location::function_name
+#ifdef _MSC_VER
+#define PRETTY_FUNCTION __FUNCSIG__
+#else
+#define PRETTY_FUNCTION __PRETTY_FUNCTION__
```

Comment:
__func__ is mostly useless in this case as it doesn't have the template arguments. I'll can do a check for gcc and clang though (they both support __PRETTY_FUNCTION__) and fall back otherwise